src/Compiler/Microsoft.AspNetCore.Razor.Language/src/BoundAttributeDescriptorBuilder.cs

```diff
@@ -1,63 +1,313 @@
+    {
+    }
 
-    public abstract RazorDiagnosticCollection Diagnostics { get; }
+    internal BoundAttributeDescriptorBuilder(TagHelperDescriptorBuilder parent, string kind)
+    {
+        _parent = parent;
+        _kind = kind;
+    }
 
+    [AllowNull]
+    public string Name { get; set; }
+    public string? TypeName { get; set; }
+    public bool IsEnum { get; set; }
+    public bool IsDictionary { get; set; }
+    public string? IndexerAttributeNamePrefix { get; set; }
+    public string? IndexerValueTypeName { get; set; }
     internal bool IsEditorRequired { get; set; }
 
-    public virtual IReadOnlyList<BoundAttributeParameterDescriptorBuilder> BoundAttributeParameters { get; }
+    public string? Documentation
+    {
+        get => _documentationObject.GetText();
+        set => _documentationObject = new(value);
+    }
+
+    public string? DisplayName { get; set; }
+
+    public IDictionary<string, string?> Metadata => _metadata.MetadataDictionary;
 
-    public virtual void BindAttributeParameter(Action<BoundAttributeParameterDescriptorBuilder> configure)
+    public void SetMetadata(MetadataCollection metadata) => _metadata.SetMetadataCollection(metadata);
+
+    public bool TryGetMetadataValue(string key, [NotNullWhen(true)] out string? value)
+        => _metadata.TryGetMetadataValue(key, out value);
+
+    public ImmutableArray<RazorDiagnostic>.Builder Diagnostics => _diagnostics ??= ImmutableArray.CreateBuilder<RazorDiagnostic>();
+
+    internal bool CaseSensitive => _parent.CaseSensitive;
+
+    public void BindAttributeParameter(Action<BoundAttributeParameterDescriptorBuilder> configure)
     {
+        if (configure == null)
+        {
+            throw new ArgumentNullException(nameof(configure));
+        }
+
+        EnsureAttributeParameterBuilders();
+
+        var builder = BoundAttributeParameterDescriptorBuilder.GetInstance(this, _kind);
+        configure(builder);
+        _attributeParameterBuilders.Add(builder);
     }
 
-#nullable enable
+    internal void SetDocumentation(string? text)
+    {
+        _documentationObject = new(text);
+    }
+
+    internal void SetDocumentation(DocumentationDescriptor? documentation)
+    {
+        _documentationObject = new(documentation);
+    }
 
-    internal virtual void SetDocumentation(string? text)
+    public BoundAttributeDescriptor Build()
     {
-        throw new NotImplementedException();
+        var diagnostics = new PooledHashSet<RazorDiagnostic>();
+        try
+        {
+            Validate(ref diagnostics);
+
+            diagnostics.UnionWith(_diagnostics);
+
+            var parameters = _attributeParameterBuilders.BuildAllOrEmpty(s_boundAttributeParameterSetPool);
+
+            var metadata = _metadata.GetMetadataCollection();
+
+            var descriptor = new BoundAttributeDescriptor(
+                _kind,
+                Name!, // Name is not expected to be null. If it is, a diagnostic will be created for it.
```

Comment:
Should we also default to `string.Empty` here?

Reply:
Yes. Jan also pointed that out. Will do that -- thanks!